src/app/outbound.rs

```diff
@@ -398,6 +398,213 @@ pub mod orig_proto_upgrade {
+        fn clone(&self) -> Self {
+            MakeSvc {
+                inner: self.inner.clone(),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    // ===== impl IdentityCheck =====
+
+    impl<M, A, B> svc::Service<http::Request<A>> for IdentityCheck<M, A, B>
+    where
+        M: svc::Service<http::Request<A>, Response = http::Response<B>>,
+        M::Error: Into<Error>,
+    {
+        type Response = M::Response;
+        type Error = Error;
+        type Future = Either<
+            FutureResult<Self::Response, Self::Error>,
+            future::MapErr<M::Future, fn(M::Error) -> Error>,
+        >;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            self.inner.poll_ready().map_err(Into::into)
+        }
+
+        fn call(&mut self, request: http::Request<A>) -> Self::Future {
+            // If the `l5d-force-id` header is present, then we should expect
+            // the target's `peer_identity` to match. If the two values do not
+            // match or there is no `peer_identity`, then we fail the request.
+            if let Some(force_identity) = identity_from_header(&request, L5D_FORCE_ID) {
+                debug!("found l5d-force-id={:?}", force_identity.as_ref());
+                match self.peer_identity {
+                    Conditional::Some(ref peer_identity) => {
+                        if force_identity != *peer_identity {
+                            warn!(
+                                "force identity check failed; found peer_identity={:?}",
+                                peer_identity
+                            );
+                            return Either::A(future::err(identity_check_error(
+                                force_identity,
+                                Some(peer_identity.clone()),
+                            )));
+                        }
+                    }
+                    Conditional::None(_) => {
+                        warn!("force identity check failed; no peer_identity found");
+                        return Either::A(future::err(identity_check_error(force_identity, None)));
+                    }
+                }
+            }
+
+            Either::B(self.inner.call(request).map_err(Into::into))
+        }
+    }
+
+    // ===== impl IdentityCheckError =====
+
+    fn identity_check_error(
+        force_identity: identity::Name,
+        peer_identity: Option<identity::Name>,
+    ) -> Error {
+        let error = IdentityCheckError {
+            force_identity,
+            peer_identity,
+        };
+
+        error.into()
+    }
+
+    impl std::error::Error for IdentityCheckError {}
+
+    impl std::fmt::Display for IdentityCheckError {
+        // TODO(kleimkuhler): Use force_identity and peer_identity fields?
+        fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+            write!(f, "force identity check failed")
+        }
+    }
+
+    impl HasH2Reason for IdentityCheckError {
```

Comment:
@hawkw If you have some time to look at this tomorrow, I had some trouble implementing `HasH2Reason` for `IdentityCheckError` to return a permission denied status code instead of a `BAD_GATEWAY` like is currently asserted in the [tests](https://github.com/linkerd/linkerd2-proxy/pull/280/files#diff-dc35718beb77b0814d17cefbd6b98593R373)